velox/type/TimestampConversion.cpp

```diff
@@ -826,6 +841,130 @@ Status parserError(const char* str, size_t len) {
+    return false;
+  }
+  auto sign = timeZoneName[start];
+  std::string_view tzView = timeZoneName;
+  std::string_view offset = tzView.substr(start + 1);
+  std::string_view hm, s;
+  bool appendZeroMinutes = false;
+  if ((offset.length() == 1 && characterIsDigit(offset[0])) ||
+      (offset.length() == 2 && allCharactersIsDigit(offset))) {
+    // h[h]
+    hm = offset;
+    // Use appendZeroMinutes to indicate that we need to append :00 to the end
+    // of timeZoneName. By appending only when assigning to timeZoneName, hm can
+    // be defined as a string_view to minimize multiple string copies.
+    appendZeroMinutes = true;
+  } else if (
+      (offset.length() == 4 && allCharactersIsDigit(offset)) ||
+      (offset.length() == 5 && allCharactersIsDigit(offset, 0, 1) &&
+       allCharactersIsDigit(offset, 3, 4))) {
+    // hh[:]mm
+    // TimeZoneMap::normalizeTimeZoneOffset will normalize hhmm to hh:mm
+    hm = offset;
+  } else if (
+      (offset.length() == 6 && allCharactersIsDigit(offset)) ||
+      (offset.length() == 8 && allCharactersIsDigit(offset, 0, 1) &&
+       allCharactersIsDigit(offset, 3, 4) &&
+       allCharactersIsDigit(offset, 6, 7))) {
+    // hh:mm:ss, hhmmss
+    auto lastSep = offset.rfind(':');
+    if (lastSep == std::string::npos && offset.length() == 6) {
+      // hhmmss
+      // TimeZoneMap::normalizeTimeZoneOffset will normalize hhmm to hh:mm.
+      hm = offset.substr(0, 4);
+      s = offset.substr(4);
+    } else if (lastSep == 5 && offset.find(':') == 2 && offset.length() == 8) {
+      // hh:mm:ss
+      hm = offset.substr(0, 5);
+      s = offset.substr(6);
+    } else {
+      return false;
+    }
+  } else {
+    return false;
+  }
+  int32_t sec = 0;
+  size_t pos = 0;
+  if (!s.empty() && !parseDoubleDigit(s.data(), s.length(), pos, sec)) {
+    return false;
+  }
+  if (sec < 0 || sec > 60) {
+    return false;
+  }
+  if (sec != 0) {
+    // Seconds timezone offset is not supported,
+    // and needs to be applied to timestamp.
+    if (sign == '+') {
+      if (ts.getSeconds() - sec < Timestamp::kMinSeconds) {
+        return false;
+      }
+      ts = Timestamp(ts.getSeconds() - sec, ts.getNanos());
+    } else if (sign == '-') {
+      if (ts.getSeconds() + sec > Timestamp::kMaxSeconds) {
+        return false;
+      }
+      ts = Timestamp(ts.getSeconds() + sec, ts.getNanos());
+    } else {
+      return false;
+    }
+  }
+  timeZoneName = sign + std::string{hm} + (appendZeroMinutes ? ":00" : "");
+  return true;
+}
+
+bool normalizeSparkTimezone(std::string& timeZoneName, Timestamp& ts) {
+  auto startsWith = [](std::string_view str, const char* prefix) {
+    return str.rfind(prefix, 0) == 0;
+  };
+  // Spark supports a timezone id with one of the prefixes UTC+, UTC-, GMT+,
+  // GMT-, UT+ or UT-, and a suffix in the following formats: h[h], hh[:]mm,
+  // hh:mm:ss, hhmmss.
+  if (startsWith(timeZoneName, "UTC") || startsWith(timeZoneName, "GMT")) {
```

Comment:
"Etc/GMT" will be handled in `TimeZoneMap`.  This code only handles some tz ids like "GMT+8", "UTC-07:06". As you mentioned earlier, Spark normalization can be added to `TimeZoneMap`, but these normalizations are specific to Spark. It might be necessary to differentiate within `TimeZoneMap` whether it's for Spark.